samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c

```diff
@@ -0,0 +1,824 @@
+ */
+static int fds_init(struct mqtt_client *c)
+{
+	if (c->transport.type == MQTT_TRANSPORT_NON_SECURE) {
+		fds.fd = c->transport.tcp.sock;
+	} else {
+#if defined(CONFIG_MQTT_LIB_TLS)
+		fds.fd = c->transport.tls.sock;
+#else
+		return -ENOTSUP;
+#endif
+	}
+
+	fds.events = POLLIN;
+
+	return 0;
+}
+
+static int do_mqtt_connect(void)
+{
+	int err = -EINVAL;
+
+	if (ctx.connected) {
+		return -EINPROGRESS;
+	}
+
+	client_init();
+
+	/* Start MQTT listening thread */
+	mqtt_thread_id = k_thread_create(&mqtt_thread,
+			mqtt_thread_stack,
+			K_THREAD_STACK_SIZEOF(mqtt_thread_stack),
+			mqtt_thread_fn, NULL, NULL, NULL,
+			THREAD_PRIORITY, 0, K_NO_WAIT);
+
+	err = mqtt_connect(&client);
+	if (err != 0) {
+		LOG_ERR("ERROR: mqtt_connect %d", err);
+		return err;
+	}
+
+	err = fds_init(&client);
+	if (err != 0) {
+		LOG_ERR("ERROR: fds_init %d", err);
+		return err;
+	}
+
+	/** start polling now for CONNACK */
+	if (err == 0) {
+		ctx.connected = true;
+		k_sem_give(&mqtt_connected);
+	}
+
+	return 0;
+}
+
+static int do_mqtt_disconnect(void)
+{
+	int err;
+
+	err = mqtt_disconnect(&client);
+	if (err) {
+		LOG_ERR("ERROR: mqtt_disconnect %d", err);
+	}
+	slm_at_mqtt_uninit();
+
+	return err;
+}
+
+static int do_mqtt_publish(u16_t qos, u16_t retain,
+				u8_t *topic, size_t topic_len,
+				u8_t *msg, size_t msg_len)
+{
+	int err = -EINVAL;
+	struct mqtt_publish_param param;
+
+	if (qos <= MQTT_QOS_2_EXACTLY_ONCE) {
+		param.message.topic.qos = (u8_t)qos;
+	} else {
+		return err;
+	}
+
+	if (retain <= 1) {
+		param.retain_flag = (u8_t)retain;
+	} else {
+		return err;
+	}
+	param.message.topic.topic.utf8 = topic;
+	param.message.topic.topic.size = topic_len;
+	param.message.payload.data = msg;
+	param.message.payload.len = msg_len;
+	param.message_id = sys_rand32_get();
+	param.dup_flag = 0;
+
+	return mqtt_publish(&client, &param);
+}
+
+static int do_mqtt_subscribe(u16_t op,
+				u8_t *topic_buf,
+				size_t topic_len,
+				u16_t qos)
+{
+	int err = -EINVAL;
+	struct mqtt_topic subscribe_topic;
+
+	const struct mqtt_subscription_list subscription_list = {
+		.list = &subscribe_topic,
+		.list_count = 1,
+		.message_id = sys_rand32_get()
+	};
+
+	if (qos <= MQTT_QOS_2_EXACTLY_ONCE) {
+		subscribe_topic.qos = (u8_t)qos;
+	} else {
+		return err;
+	}
+	subscribe_topic.topic.utf8 = topic_buf;
+	subscribe_topic.topic.size = topic_len;
+
+	if (op == 1) {
+		err = mqtt_subscribe(&client, &subscription_list);
+	} else if (op == 0) {
+		err = mqtt_unsubscribe(&client, &subscription_list);
+	}
+
+	return err;
+}
+
+/**@brief handle AT#XMQTTCON commands
+ *  AT#XMQTTCON=<op>[,<url>,<port>[,<sec_tag>]]
+ *  AT#XMQTTCON?
+ *  AT#XMQTTCON=?
+ */
+static int handle_at_mqtt_connect(enum at_cmd_type cmd_type)
+{
+	int err = -EINVAL;
+
+	u16_t op;
+	size_t url_sz = MQTT_MAX_URL_LEN;
+	size_t cid_sz = MQTT_MAX_CID_LEN;
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		if (at_params_valid_count_get(&at_param_list) == 0) {
```

Comment:
No, the count starts from AT command itself, which is index 0, <op> is index 1. So the check should be count > 1

Reply:
As discussed, keep using strict check.